src/docs/development/platform-integration/add-to-app-android/add-flutter-fragment.md

```diff
@@ -0,0 +1,311 @@
+    }
+}
+```
+
+The above code is sufficient to render a Flutter UI that begins with a call to
+your `main()` Dart entrypoint and an initial Flutter route of `/`. However, this
+code is not sufficient to achieve all expected Flutter behavior. Flutter depends
+on various OS signals that need to be forwarded from your host `Activity` to
+`FlutterFragment`. These calls are shown below:
+
+```java
+public class MyActivity extends FragmentActivity {
+    @Override
+    public void onPostResume() {
+        super.onPostResume();
+        flutterFragment.onPostResume();
+    }
+
+    @Override
+    protected void onNewIntent(@NonNull Intent intent) {
+        flutterFragment.onNewIntent(intent);
+    }
+
+    @Override
+    public void onBackPressed() {
+        flutterFragment.onBackPressed();
+    }
+
+    @Override
+    public void onRequestPermissionsResult(
+        int requestCode, 
+        @NonNull String[] permissions, 
+        @NonNull int[] grantResults
+    ) {
+        flutterFragment.onRequestPermissionsResult(
+            requestCode, 
+            permissions, 
+            grantResults
+        );
+    }
+
+    @Override
+    public void onUserLeaveHint() {
+        flutterFragment.onUserLeaveHint();
+    }
+
+    @Override
+    public void onTrimMemory(int level) {
+        super.onTrimMemory(level);
+        flutterFragment.onTrimMemory(level);
+    }
+}
+```
+
+With the above OS signals forwarded to Flutter, your `FlutterFragment` works as
+expected. You have now added a `FlutterFragment` to your existing Android app.
+
+## Using a pre-warmed `FlutterEngine`
+
+By default, a `FlutterFragment` creates its own instance of a `FlutterEngine`,
+which requires non-trivial warmup time. This means your user sees a blank
+`Fragment` for a second or two. You can mitigate most of this warmup time by
+using an existing, pre-warmed instance of `FlutterEngine`.
+
+Please see the [instructions for instantiating and starting a `FlutterEngine`].
+
+[instructions for instantiating and starting a `FlutterEngine`]: /docs/development/platform-integrations/add-to-app-android/flutter-engine
+
+To use a pre-warmed `FlutterEngine` in a `FlutterFragment`, subclass
+`FlutterFragment` and override `provideFlutterEngine()` as shown below:
+
+```java
+public class MyFlutterFragment extends FlutterFragment {
+    @Override
+    protected FlutterEngine provideFlutterEngine(@NonNull Context context) {
+        // Obtain a reference to, and return, your pre-warmed FlutterEngine.
+        // For example:
+        return ((MyApplication) context.getApplication())
+            .getPreWarmedFlutterEngine();
+    }
+
+    // When providing a pre-warmed engine within a FlutterFragment subclass,
+    // be sure to override this method and return "true" so that this
+    // FlutterFragment does not destroy your FlutterEngine when the Fragment
+    // is destroyed.
+    @Override
+    protected boolean retainFlutterEngineAfterFragmentDestruction() {
+        return true;
```

Comment:
When should this be false when provideFlutterEngine returns an engine that isn't one that the super class built by default?

Reply:
Yeah, it should be false when no engine is provided, otherwise there could be a memory leak. But I suppose there could also be a case where a developer provides a FlutterEngine and still wants it to be destroyed when the Fragment is destroyed.